pkg/reconciler/revision/revision.go

```diff
@@ -144,17 +144,31 @@ func (c *Reconciler) reconcileDigest(ctx context.Context, rev *v1alpha1.Revision
 		ServiceAccountName: rev.Spec.ServiceAccountName,
 		ImagePullSecrets:   imagePullSecrets,
 	}
-	digest, err := c.resolver.Resolve(rev.Spec.GetContainer().Image,
-		opt, cfgs.Deployment.RegistriesSkippingTagResolving)
-	if err != nil {
-		err = fmt.Errorf("failed to resolve image to digest: %w", err)
-		rev.Status.MarkContainerHealthyFalse(v1alpha1.ContainerMissing,
-			v1alpha1.RevisionContainerMissingMessage(
-				rev.Spec.GetContainer().Image, err.Error()))
-		return err
-	}
 
-	rev.Status.ImageDigest = digest
+	for i := range rev.Spec.Containers {
+		digest, err := c.resolver.Resolve(rev.Spec.Containers[i].Image,
+			opt, cfgs.Deployment.RegistriesSkippingTagResolving)
+		if err != nil {
+			err = fmt.Errorf("failed to resolve image to digest: %w", err)
+			rev.Status.MarkContainerHealthyFalse(v1alpha1.ContainerMissing,
+				v1alpha1.RevisionContainerMissingMessage(
+					rev.Spec.Containers[i].Image, err.Error()))
+			return err
+		}
+		// digest assignment based on number of containers
+		if len(rev.Spec.Containers) == 1 {
+			rev.Status.ImageDigest = digest
+		} else if len(rev.Spec.Containers) > 1 {
+			if len(rev.Spec.Containers[i].Ports) != 0 {
+				rev.Status.ImageDigest = digest
+			} else {
+				if rev.Status.ImageDigests == nil {
+					rev.Status.ImageDigests = make(map[string]string)
+				}
+				rev.Status.ImageDigests[rev.Spec.Containers[i].Name] = digest
```

Comment:
I wonder if we should make this "image-as-passed -> digest" rather than "containerName -> digest". That'd even deduplicate things on the degenerate case of having one image used multiple times. We should probably also only resolve those once to not have a race if you've **juuuust** pushed an update as we've been resolving and to guarantee the same version.